tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,390 @@
+	{
+		Form_continuous_aggs_materialization_invalidation_log form;
+		CatalogSecurityContext sec_ctx;
+		MemoryContext oldmctx;
+		TupleInfo *ti;
+		bool should_free;
+		HeapTuple tuple;
+		HeapTuple update_tuple = NULL;
+
+		oldmctx = MemoryContextSwitchTo(state->per_tuple_mctx);
+		ti = ts_scan_iterator_tuple_info(&iterator);
+		tuple = ts_scan_iterator_fetch_heap_tuple(&iterator, false, &should_free);
+		form = (Form_continuous_aggs_materialization_invalidation_log) GETSTRUCT(tuple);
+
+		/* Entry is completely enclosed by the refresh window */
+		if (form->lowest_modified_value >= refresh_window->start &&
+			form->greatest_modified_value < refresh_window->end)
+		{
+			/*
+			 * Entry completely enclosed so can be deleted:
+			 *
+			 * |---------------|
+			 *     [+++++]
+			 */
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_delete_tid_only(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+			ts_catalog_restore_user(&sec_ctx);
+		}
+		else if (form->lowest_modified_value < refresh_window->start &&
+				 form->greatest_modified_value >= refresh_window->end)
+		{
+			HeapTuple newtup;
+			/*
+			 * Need to cut in both ends:
+			 *
+			 *     |------|
+			 * [+++++++++++++]
+			 *
+			 * [++]        [+]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   form->lowest_modified_value,
+												   refresh_window->end - 1);
+			newtup =
+				create_invalidation_tup(tuple, form->greatest_modified_value, refresh_window->end);
+
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_insert_only(ti->scanrel, newtup);
+			ts_catalog_restore_user(&sec_ctx);
+			heap_freetuple(newtup);
+		}
+		else if (form->lowest_modified_value < refresh_window->start &&
+				 form->greatest_modified_value >= refresh_window->start &&
+				 form->greatest_modified_value < refresh_window->end)
+		{
+			/*
+			 * Need to cut in right end:
+			 *
+			 *     |------|
+			 * [++++++]
+			 *
+			 * [++]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   form->lowest_modified_value,
+												   refresh_window->end - 1);
+		}
+		else if (form->lowest_modified_value >= refresh_window->start &&
+				 form->lowest_modified_value < refresh_window->end &&
+				 form->greatest_modified_value >= refresh_window->end)
+		{
+			/*
+			 * Need to cut in left end:
+			 *
+			 * |------|
+			 *    [++++++++]
+			 *
+			 *         [+++]
+			 */
+			update_tuple = create_invalidation_tup(tuple,
+												   refresh_window->start,
+												   form->greatest_modified_value);
+		}
+
+		if (update_tuple)
+		{
+			ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+			ts_catalog_update_tid_only(ti->scanrel, ts_scanner_get_tuple_tid(ti), update_tuple);
+			ts_catalog_restore_user(&sec_ctx);
+			heap_freetuple(update_tuple);
+		}
+		MemoryContextSwitchTo(oldmctx);
+		MemoryContextReset(state->per_tuple_mctx);
+	}
+
```

Comment:
another potential optimization:  if we order the entries, we can collapse multiple entries into 1 if there are overlaps.